Cutelyst/Plugins/Utils/Validator/validatordomain.cpp

```diff
@@ -0,0 +1,306 @@
+                                                            valid = false;
+                                                            diag = InvalidChars;
+                                                            break;
+                                                        }
+                                                    } else {
+                                                        if (isPunyCode) {
+                                                            if (!(isDigit || isDash || isChar)) {
+                                                                valid = false;
+                                                                diag = InvalidTLD;
+                                                                break;
+                                                            }
+                                                        } else {
+                                                            if (!isChar) {
+                                                                valid = false;
+                                                                diag = InvalidTLD;
+                                                                break;
+                                                            }
+                                                        }
+                                                    }
+                                                }
+                                            } else {
+                                                valid = false;
+                                                diag = LabelTooLong;
+                                                break;
+                                            }
+                                        } else {
+                                            valid = false;
+                                            diag = EmptyLabel;
+                                            break;
+                                        }
+                                    } else {
+                                        break;
+                                    }
+                                }
+                            } else {
+                                valid = false;
+                                diag = InvalidTLD;
+                            }
+                        } else {
+                            valid = false;
+                            diag = InvalidLabelCount;
+                        }
+                    } else {
+                        valid = false;
+                        diag = TooLong;
+                    }
+                } else {
+                    valid = false;
+                    diag = EmptyLabel;
+                }
+            }
+        } else {
+            valid = false;
+            diag = EmptyLabel;
+        }
+    } else {
+        valid = false;
+        diag = EmptyLabel;
+    }
+
+
+    if (valid && checkDNS) {
```

Comment:
I think it would be interesting to have a non-blocking check, even if you create a local event loop here.